server/src/main/java/com/linecorp/centraldogma/server/internal/replication/ZooKeeperCommandExecutor.java

```diff
@@ -714,33 +751,89 @@ public void childEvent(CuratorFramework unused, PathChildrenCacheEvent event) th
+                executionPath, k -> new InterProcessMutex(curator, absolutePath(LOCK_PATH, k)));
 
+        WriteLock writeLock = null;
         try {
             mtx.acquire();
+            if (command instanceof PushCommand) {
+                writeLock = acquireWriteLock((PushCommand) command);
+            }
         } catch (Exception e) {
             logger.error("Failed to acquire a lock for {}; entering read-only mode", executionPath, e);
             stopLater();
             throw new ReplicationException("failed to acquire a lock for " + executionPath, e);
         }
 
-        return () -> {
-            try {
-                mtx.release();
-            } catch (Exception ignored) {
-                // Ignore.
-            }
-        };
+        if (writeLock != null) {
+            releaseWriteLock(writeLock, mtx, executionPath);
+        }
+
+        return () -> safeRelease(mtx);
+    }
+
+    private static void safeRelease(InterProcessMutex mtx) {
+        try {
+            mtx.release();
+        } catch (Exception ignored) {
+            // Ignore.
+        }
+    }
+
+    @Nullable
+    private WriteLock acquireWriteLock(PushCommand command) throws Exception {
+        if (command.projectName().equals(INTERNAL_PROJ) ||
+            command.repositoryName().equals(Project.REPO_DOGMA)) {
+            // Do not check quota for internal project and repository.
+            return null;
+        }
+
+        final RepositoryMetadata meta;
+        try {
+            meta = metadataService.getRepo(command.projectName(), command.repositoryName()).get();
+        } catch (InterruptedException | ExecutionException e) {
+            throw new CentralDogmaException("Unexpected exception caught while retrieving " +
+                                            RepositoryMetadata.class.getSimpleName(), e);
+        }
+
+        QuotaConfig writeQuota = meta.writeQuota();
+        if (writeQuota == null) {
+            writeQuota = this.writeQuota;
+        }
+        if (writeQuota == null) {
+            return null;
+        }
+
+        final int requestQuota = writeQuota.requestQuota();
+        final Entry<InterProcessSemaphoreV2, SettableSharedCount> semaphoreAndCount =
+                semaphoreMap.computeIfAbsent(command.executionPath(), k -> {
+                    final SettableSharedCount cnt = new SettableSharedCount(requestQuota);
+                    return new SimpleImmutableEntry<>(
+                            new InterProcessSemaphoreV2(curator, absolutePath(QUOTA_PATH, k), cnt), cnt);
+                });
+
+        final SettableSharedCount count = semaphoreAndCount.getValue();
+        if (count.getCount() != requestQuota) {
+            count.setCount(requestQuota);
+        }
+        final InterProcessSemaphoreV2 semaphore = semaphoreAndCount.getKey();
+        final Lease lease = semaphore.acquire(200, TimeUnit.MILLISECONDS);
```

Comment:
There is no `tryAcquire()` style method in semaphore. It takes time to require a lock because of network communication even though it is available now.
We need a sensible default for this. 🤔

Reply:
Yeah, we need that. Do you know how long does it take to acquire a lease in the multi-datacenter environment?

Reply:
> Do you know how long does it take to acquire a lease in the multi-datacenter environment?

I didn't measure only a locking time. We need a test for it. 
When I ran a performance test with 3 multi-datacenter, the average QPS was 7~8 per repository.
So I thought 200ms is enough time to work Central Dogma normally.

Reply:
> the average QPS was 7~8 per repository.
This QPS test was executed including this lease mechanism, right?
Then, 200ms will be fine I guess. 😄